Issue 1 ユーザー情報登録画面の作成

qutti96 opened the pull request:
# 作成内容
userInfo.vue(ユーザー情報登録画面)  firestore登録処理を作成
## できなかったこと
createdAtの時間登録処理にエラーがでる

`            createdAt: db.FieldValue.serverTimestamp(),
`
>  client.js?06a0:96 TypeError: Cannot read property 'serverTimestamp' of undefined

## 未実装
お子さま二人目追加とその処理

qutti96 committed:
不要な記述の削除

qutti96 committed:
user.idの表示

qutti96 committed:
ユーザー情報登録画面作成・firestore登録処理（createedAt以外）

version-1 commented on pages/userSignup.vue:
```diff
@@ -170,4 +211,81 @@
+        createdAt: '',
+        updateAt: '',
+      },
+      signupFormValidation: {
+        valid: false,
+        nameRules: [(v) => !!v || '名前は必須項目です'],
+        telRules: [
+          (v) => !!v || '電話番号は必須項目です',
+          (v) => /[0-9]{11}/.test(v) || '半角数字11文字で入力してください',
+        ],
+        postalcodeRules: [
+          (v) => !!v || '郵便番号は必須項目です',
+          (v) => /[0-9]{7}/.test(v) || '半角数字7文字で入力してください',
+        ],
+        addressRules: [(v) => !!v || '住所は必須項目です'],
+        emailRules: [
+          (v) => !!v || 'メールアドレスは必須項目です',
+          (v) => /.+@.+/.test(v) || '正しいE-mail形式でご記入ください',
+        ],
+        childRules: [(v) => !!v || 'お子さま一人目の入力は必須項目です'],
+      },
```
書き換える必要のないものは、`export default`の上に書いてしまった方がよいですかね。

version-1 replied:
```javascript
const signupFormRules = {
        name: [(v) => !!v || '名前は必須項目です'],
        tel: [
          (v) => !!v || '電話番号は必須項目です',
          (v) => /[0-9]{11}/.test(v) || '半角数字11文字で入力してください',
        ],
        postalcode: [
          (v) => !!v || '郵便番号は必須項目です',
          (v) => /[0-9]{7}/.test(v) || '半角数字7文字で入力してください',
        ],
        address: [(v) => !!v || '住所は必須項目です'],
        email: [
          (v) => !!v || 'メールアドレスは必須項目です',
          (v) => /.+@.+/.test(v) || '正しいE-mail形式でご記入ください',
        ],
        child: [(v) => !!v || 'お子さま一人目の入力は必須項目です'],
}

export default {
   ...
}
```

version-1 replied:
```jsx
<v-text-field
                  v-model="email"
                  :rules="emailRules"
                  v-model="signupForm.uchild.child1"
                  :rules="signupFormRules.child"
                  label="お子さま一人目のお名前"
                  placeholder="太郎"
                  :counter="10"
                  maxlength="10"
                  clearable
                  required
                ></v-text-field>
```

qutti96 replied:
書き直してみました。
https://github.com/qutti96/care-log-app/pull/2/commits/5ac5a24d5f11ba2f89a19c40e448f87da1273762
が、エラーが出ました

```javascript
webpack-internal:///./node_modules/vue/dist/vue.runtime.esm.js:620 
[Vue warn]: Property or method "signupFormRules" is not defined
 on the instance but referenced during render. 
Make sure that this property is reactive, either in the data option,
 or for class-based components, by initializing the property. 
See: https://vuejs.org/v2/guide/reactivity.html#Declaring-Reactive-Properties.
```

version-1 commented on pages/userSignup.vue:
```diff
@@ -170,4 +211,81 @@
+      this.$refs.signupForm.validate()
+    },
+    addUser() {
+      if (this.$refs.signupForm.validate()) {
```
上でmethodsを定義しているので
`if (this.validate())` でいけるかと思います。

computedに定義するのであれば
`if (this.validate)` ですね。

version-1 replied:
早期リターンしてあげるとコードが見やすくなりますかね。

```javascript
if (!this.validate) {
   return 
}

const date = this.$date(new Date(), 'DD MMMM, YYYY')
...
```

qutti96 replied:
```javascript
if (!this.validate) {
   return 
}
```
この記述にしたところ、登録するボタンを押すと
必須項目が入力されないまま、フィールドがからの状態で新ドキュメントが追加されてしまったので現状のままにしました。

qutti96 replied:
```javascript
if (!this.validate) {
}
```
実装しました。
https://github.com/qutti96/care-log-app/commit/b6c226160f3dfbdd2eaeb64089addeb628141402

version-1 commented on pages/userSignup.vue:
```diff
@@ -170,4 +211,81 @@
+    }
+  },
+  methods: {
+    validate() {
+      this.$refs.signupForm.validate()
+    },
```
データを書き換えない処理なので、computedに書いてあげると良いと思います。

qutti96 replied:
computedに記述したらエラーが出たのでそのままにしました。

`client.js?06a0:96 TypeError: this.validate is not a function`

https://vuetifyjs.com/en/components/forms/#misc
vuetifyの仕様？ということでこのままでいいでしょうか、、

version-1 replied:
`this.validate()`

`this.validate`

version-1 replied:
```javascript
validate() {
      return this.$refs.signupForm.validate()
    }
```

qutti96 replied:
解決しました
24f071e

ありがとうございます。

version-1 commented on pages/userSignup.vue:
```diff
@@ -170,4 +211,81 @@
+    addUser() {
+      if (this.$refs.signupForm.validate()) {
+        const date = this.$date(new Date(), 'DD MMMM, YYYY')
+        db.collection('users')
+          .add({
+            ufirstname: this.signupForm.ufirstname,
+            ulastname: this.signupForm.ulastname,
+            ufirstkana: this.signupForm.ufirstkana,
+            ulastkana: this.signupForm.ulastkana,
+            uid: this.signupForm.uid,
+            utel: this.signupForm.utel,
+            uemail: this.signupForm.uemail,
+            upostalcode: this.signupForm.upostalcode,
+            uaddress: this.signupForm.uaddress,
+            uchild: {
+              child1: this.signupForm.uchild.child1,
+            },
+            // createdAt: date,
+            createdAt: db.FieldValue.serverTimestamp(),
+          })
```
signupFormの内容を保存する形かと思うのでスプレッド演算子を使うと良いと思います。

```javascript
 db.collection('users')
          .add({
            ...this.singupForm,
            // createdAt: date,
            createdAt: db.FieldValue.serverTimestamp(),
          })
```

ですかね。

version-1 replied:
```javascript
createdAt: firebase.firestore.FieldValue.serverTimestamp()
```
で試してみると良いかもしれません。

qutti96 replied:
```javascript
createdAt: firebase.firestore.FieldValue.serverTimestamp()
```
実装しコミットしました。
https://github.com/qutti96/care-log-app/pull/2/commits/ee1dbbb35afe0bdb0e76518aa8deb88efc6bd918
スプレッド演算子
f4ac20d

version-1 commented on pages/userSignup.vue:
```diff
@@ -170,4 +211,81 @@
+    validate() {
+      this.$refs.signupForm.validate()
+    },
+    addUser() {
```
可能であれあばasync/awaitでかけると良いと思います。

qutti96 replied:
https://github.com/firebase/snippets-node/blob/7218a08861c3bbb6ff207315ddf9030da1c7f2c6/firestore/main/index.js#L174-L180
上記URL参考にして書いてリファクトしました。
3ad9d54

処理は通りましたが、これでいいのかよくわかっていません、、

version-1 commented:
https://jp.vuejs.org/v2/style-guide/index.html#%E5%8D%98%E4%B8%80%E3%82%A4%E3%83%B3%E3%82%B9%E3%82%BF%E3%83%B3%E3%82%B9%E3%81%AE%E3%82%B3%E3%83%B3%E3%83%9D%E3%83%BC%E3%83%8D%E3%83%B3%E3%83%88%E5%90%8D-%E5%BC%B7%E3%81%8F%E6%8E%A8%E5%A5%A8

qutti96 committed:
フォームボタン・リセット処理追加・createdAtフィールド追加

qutti96 committed:
不要記述の削除 #2

qutti96 committed:
ログイン画面(login.vue) 作成 #3

qutti96 committed:
issue_1からcheckoout

qutti96 committed:
バリデーションルールをexport dataから外に出す #2

qutti96 committed:
dotenv 導入

qutti96 committed:
return 追記  | Issue 3 ログイン画面(login.vue) 作成 #3

qutti96 committed:
不要記述削除  | Issue 3 ログイン画面(login.vue) 作成 #3

qutti96 committed:
ブランチ issue_3からマージ
Merge commit 'd387201f251bcdacc1a6922fbf20410223a2bb0a' into issue_1

# Conflicts:
#	pages/userSignup.vue
#	plugins/firebase.js

qutti96 committed:
ブランチissue_3からマージ
Merge commit '927fcd7f8beea26b30c928f0cd81eb4e806e3e15' into issue_1

qutti96 committed:
ブランチissue_3からマージ
Merge commit '927fcd7f8beea26b30c928f0cd81eb4e806e3e15' into issue_1	02662a4	qutti96 <qutti96@gmail.com>	2020/10/14 14:27

Merge branch 'issue_3' into issue_1

qutti96 committed:
バリデーション機能にreturn追記 | Issue 1 ユーザー情報登録画面の作成 #2

qutti96 committed:
リファクト：スプレッド演算子を使う  | Issue 1 ユーザー情報登録画面の作成 #1 #2

qutti96 committed:
リファクト：async/await 処理に変更  | Issue 1 ユーザー情報登録画面の作成 #1 #2

qutti96 committed:
リファクト：バリデート処理をcomputedに記述  | Issue 1 ユーザー情報登録画面の作成 #1 #2
